### ISSUE TITLE
mmdetectionv2 GPU memory have a high occupancy

### ISSUE THREAD

**Issue by oysz2016** [open-mmlab/mmdetection#2993]
I used the mask RCNN +R50+FPN structure. When I used V1.0 version, GPU memory would take up 2-3GB, but v2 version would take up 7G. May I ask why this might be caused.

**Comment by yhcao6** [comment]
The memory usage in mmdetv2.0 is indeed a little higher than mmdetv1.0, but the gap should not be so large.
Are you using your own dataset? Could you provide us more detailed information?

**Comment by oysz2016** [comment]
Yes, I use my own data set, the memory footprint seems to be related to the number of targets detected in the image.	My environmental information is as follows:
sys.platform: linux
Python: 3.6.10 |Anaconda, Inc.| (default, May  8 2020, 02:54:21) [GCC 7.3.0]
CUDA available: True
CUDA_HOME: /usr/local/cuda-10.0
NVCC:
GPU 0: Tesla V100-SXM2-16GB
GCC: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
PyTorch: 1.3.1
PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - Intel(R) Math Kernel Library Version 2020.0.1 Product Build 20200208 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v0.20.5 (Git Hash 0125f28c61c1f822fd48570b4c1066f96fcb9b2e)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CUDA Runtime 10.0
  - NVCC architecture flags: -gencode;arch=compute_35,code=sm_35;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_50,code=compute_50
  - CuDNN 7.6.5
    - Built with CuDNN 7.6.4
  - Magma 2.5.0
  - Build settings: BLAS=MKL, BUILD_NAMEDTENSOR=OFF, BUILD_TYPE=Release, CXX_FLAGS=-fvisibility-inlines-hidden -std=c++11 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -I/opt/conda/conda-bld/pytorch_1574381331675/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place/include -fdebug-prefix-map=/opt/conda/conda-bld/pytorch_1574381331675/work=/usr/local/src/conda/pytorch-1.3.1 -fdebug-prefix-map=/opt/conda/conda-bld/pytorch_1574381331675/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_place=/usr/local/src/conda-prefix -Wno-deprecated -fvisibility-inlines-hidden -fopenmp -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Wno-stringop-overflow, DISABLE_NUMA=1, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, USE_CUDA=True, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, USE_STATIC_DISPATCH=OFF,

TorchVision: 0.4.2
OpenCV: 4.2.0
MMCV: 0.5.9
MMDetection: 2.1.0+e378675
MMDetection Compiler: GCC 7.5
MMDetection CUDA Compiler: 10.0

**Comment by yhcao6** [comment]
Yes, I guess that is the main reason of large memory usage. In MMDet1.x, mask target computation is on CPU 
https://github.com/open-mmlab/mmdetection/blob/25ede58af6072ea8089af7b42cbb418a3b6991e9/mmdet/core/mask/mask_target.py#L16-L41, 
but in MMDet2.x the computation is on GPU
https://github.com/open-mmlab/mmdetection/blob/e7537a4b394728eb02133e27b5aadcd8caec8268/mmdet/core/mask/structures.py#L209-L254
Maybe you can try PolygonMask, you can just set `poly2mask=False` in config
https://github.com/open-mmlab/mmdetection/blob/e7537a4b394728eb02133e27b5aadcd8caec8268/configs/mask_rcnn/mask_rcnn_r50_fpn_poly_1x_coco.py#L9-L15

**Comment by oysz2016** [comment]
> Yes, I guess that is the main reason of large memory usage. In MMDet1.x, mask target computation is on CPU
> https://github.com/open-mmlab/mmdetection/blob/25ede58af6072ea8089af7b42cbb418a3b6991e9/mmdet/core/mask/mask_target.py#L16-L41
> 
> ,
> but in MMDet2.x the computation is on GPU
> https://github.com/open-mmlab/mmdetection/blob/e7537a4b394728eb02133e27b5aadcd8caec8268/mmdet/core/mask/structures.py#L209-L254
> 
> 
> Maybe you can try PolygonMask, you can just set `poly2mask=False` in config
> https://github.com/open-mmlab/mmdetection/blob/e7537a4b394728eb02133e27b5aadcd8caec8268/configs/mask_rcnn/mask_rcnn_r50_fpn_poly_1x_coco.py#L9-L15

But is there such a big difference? Strangely, I looped two images separately, and the memory was 4GB. But When I loop it all together, it takes up to 8GB of GPU memory, and it works just fine with 4GB

**Comment by yhcao6** [comment]
Yes, the gap will indeed be so big. I hack the `RCNN assigner` in `mask rcnn` config to downscale the `pos_iou` threshold to be 0, the portion is positive set to be 1, so that all boxes are positive samples. And test with 1 GPU the memory usage increase around 4GB.

**Comment by zytx121** [comment]
> Yes, the gap will indeed be so big. I hack the `RCNN assigner` in `mask rcnn` config to downscale the `pos_iou` threshold to be 0, the portion is positive set to be 1, so that all boxes are positive samples. And test with 1 GPU the memory usage increase around 4GB.

When i set `poly2mask=False` in config, the training process is stuck. But when I set it to `True`, I can train normally, despite the high memory usage. Is something wroing here?
